nanocoat/lib/scritchui/scritchPencilRasterChunky.c

```diff
@@ -84,22 +84,111 @@ sjme_errorCode sjme_scritchpen_core_copyArea(
 	sjme_attrInValue sjme_jint anchor)
 {
 	sjme_errorCode error;
+	sjme_jint tmpSx, tmpSy, scanlineSize;
+	sjme_cpointer copiedArea;
 	
 	if (g == NULL)
 		return SJME_ERROR_NULL_ARGUMENTS;
+
+	/* The source rectangle must always be in bounds. */
+	if (sx < 0 || sx < 0 || sx + w > g->width || sy + h > g->height)
+		return SJME_ERROR_INDEX_OUT_OF_BOUNDS;
+
+	/* Drawing nothing? */
+	if (w <= 0 || h <= 0)
+		return SJME_ERROR_NONE;
 	
 	/* Need to lock? */
 	if (sjme_error_is(error = sjme_scritchpen_core_lock(g)))
 		return sjme_error_default(error);
 	
-	sjme_todo("Impl?");
-	return sjme_error_notImplemented(0);
+	error = SJME_ERROR_NONE;
+
+	/* Those are needed to translate back before moving to the dest area. */
+	tmpSx = -sx;
+	tmpSy = -sy;
+
+	/* Translate base coordinates. */
+	sjme_scritchpen_coreUtil_applyTranslate(g, &sx, &sy);
+
+	/**
+	 * A neat property here is that we don't need to worry about format
+	 * conversion, as the data is being copied between regions of the same
+	 * image. We can get away with just figuring out the native image format,
+	 * and then go straight to copying from there.
+	 */
+	scanlineSize = (w * g->bitsPerPixel) / 8;
+	copiedArea = sjme_alloca(scanlineSize * h);
+	if (copiedArea == NULL)
+		return sjme_error_defaultOr(error,
+			sjme_error_outOfMemory(NULL, scanlineSize * h));
 	
+	memset(copiedArea, 0, scanlineSize * h);
+
+	for(int y = 0; y < h; y++)
+	{
+		error |= g->prim.rawScanGet(g, sx, sy + y,
+			(copiedArea + y * scanlineSize), scanlineSize, w);
+	}
+
+	sjme_scritchpen_coreUtil_applyTranslate(g, &tmpSx, &tmpSy);
+
+	/* Now start moving to the destination in order to copy into it. */
+
+	/* Apply anchor to the destination area. */
+	error = sjme_scritchpen_coreUtil_applyAnchor(anchor, dx, dy, w, h, 0, &dx,
+		&dy);
+
+	sjme_scritchpen_coreUtil_applyTranslate(g, &dx, &dy);
+
+	/**
+	 * The destination area on the other hand, can extend beyond the context
+	 * bounds, we just don't copy whatever's outside.
+	 */
+	if (dx < 0)
+	{
+		w += dx;
+		dx = 0;
+	}
+	if (dy < 0)
+	{
+		h += dy;
+		dy = 0;
+	}
+	if(dx + w > g->width)
+		w = g->width - dx;
+	if(dy + h > g->height)
+		h = g->height - dy;
+
+	/* Drawing nothing? */
+	if (w <= 0 || h <= 0)
+		goto nodraw;
```

Comment:
This can be deleted as the condition is already checked for at the start, thus no other code nor allocations take place.